src/System/Console/Hawk.hs

```diff
@@ -50,8 +53,16 @@ processSpec :: HawkSpec -> IO ()
 processSpec Help          = help
 processSpec Version       = putStrLn versionString
 processSpec (Eval  e   o) = myRunUncertainIO e $ processEvalSpec  (contextSpec e)   o (userExpr e)
-processSpec (Apply e i o) = myRunUncertainIO e $ processApplySpec (contextSpec e) i o (userExpr e)
-processSpec (Map   e i o) = myRunUncertainIO e $ processMapSpec   (contextSpec e) i o (userExpr e)
+processSpec (Apply e i o) = myRunUncertainIO e (processApplySpec (contextSpec e) i o (userExpr e)) >> overwrite o
+processSpec (Map   e i o) = myRunUncertainIO e (processMapSpec   (contextSpec e) i o (userExpr e)) >> overwrite o
+
+-- | Overwrites the input file with the temporary output file if it exists
+overwrite :: OutputSpec -> IO ()
+overwrite (OutputSpec (OutputFile (FileSink f _)) _) = do
+    fe <- doesFileExist tmpFile
+    when fe $ renameFile tmpFile f
```

Comment:
Under which circumstances would the temporary file not exist? If that happens, wouldn't we prefer to let the user know by failing loudly rather than silently doing nothing?

Reply:
Now that I am taking a closer look, I see that it could happen if the `UncertainT` computation ends with an error. In that case, I agree that we don't want to try to perform an overwrite, but I'd rather use `UncertainT`'s monadic short-circuiting to skip over the `overwrite` than to check whether we've reached the point in the logic where the temporary file has been created. After all, what if the error occurs after the temporary file has been created but before all the contents is written to it? Let's move the `overwrite` call inside the `UncertainT` computation.

Reply:
Or even better, move the `overwrite` logic inside `outputRows`

Reply:
> Or even better, move the overwrite logic inside outputRows

I can't move this logic into outputRows because of lazy input reading. The input file is blocked until we force evaluation which only happens with strict reading.

Reply:
I don't understand why laziness would cause any issues here. Lazy reading means that the input file won't be closed until its bytestring has been completely consumed. In `outputRows`, the `B.writeFile (f <.> tmp) s` should completely consume the `s`, so the file should be available immediately after that.

Reply:
Oh, but this `s` is the output bytestring, not the input bytestring! If the user expression only looks at part of the input, e.g. `L.take 10`, consuming the entire output bytestring won't cause the entire input bytestring to be consumed, so the input file won't be closed.

But why is the input file now considered closed after the `myRunUncertainIO` block? That seems fishy. I would prefer to make sure the file is closed by manually calling `hClose` on it. Perhaps we could modify `getInputString` so it would also return an IO action which ensures the input is closed? We could then pass this action to `outputRows`, which would call the action immediately after the `B.writeFile`.

This in-place feature has more subtleties than I expected!